lib/tokenrange.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * Cppcheck - A tool for static C/C++ code analysis
+ * Copyright (C) 2007-2020 Cppcheck team.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ //---------------------------------------------------------------------------
+#ifndef tokenrangeH
+#define tokenrangeH
+//---------------------------------------------------------------------------
+
+#include "config.h"
+
+template<typename T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)>
+class TokenRangeBase
+{
+    T* mFront;
+    T* mBack;
+
+public:
+    TokenRangeBase(T* front, T* back) : mFront(front), mBack(back) {}
+
+    struct TokenIterator
+    {
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = T*;
+        using difference_type = std::ptrdiff_t;
+        using pointer = void;
```

Comment:
It is common for things like output iterators to set the typedef to `void` when its not applicable. You can see that for [`std::back_insert_iterator`](https://en.cppreference.com/w/cpp/iterator/back_insert_iterator) as an example. Since the token iterator doesn't define `operator->`, then it seems to make sense to set this to `void` as well.